File: Fortran/ex2/costFunctionReg.f90
	subroutine costFunctionReg(M, N, X, y, theta, JCost, grad)
	implicit none
	integer::i,j,  M, N
	double precision,dimension(M,N) :: X
	double precision,dimension(M,1) :: y
	double precision,dimension(N,1) :: theta, grad
	double precision :: JCost
	double precision :: sigmoid,  g, lambda, TX, theta2
	
	lambda = 1.
	
	theta2 = 0
	grad = 0
	JCost =0
	
	
	do i = 1, M
		   TX =0
		 do j = 1, N
		   TX = TX + theta(j,1)*X(i, j)
		 enddo
		g = sigmoid( TX)		
				 
		JCost = JCost - y(i,1) * log(g)- (1- y(i,1))* log(1- g)
		
		do j = 1, N
		   grad(j,1) = grad(j,1) + X(i,j)* (g - y(i,1))
		enddo
	enddo	
	
		do j = 1, N
		   theta2 = theta2 + theta(j, 1)**2
		 enddo
		 
	  JCost = JCost/m + lambda/2/m *theta2
	  
	           grad = grad/m
		do j = 1, N
		      grad(j,1) = grad(j, 1) + lambda /m *theta(j, 1)
		enddo
	end

File: C/ex2/costFunctionReg.c
#include <stdio.h>
#include <math.h>
double sigmoid(double z);

double costFunctionReg(int M, int N, double theta[N], double X[M][N], double y[M], double grad[N], double lambda)
{
	int i, j ;
	double   g, JCost,  TX=0, theta2;
	
	for(j=0; j < N; j++){
	        grad[j] = 0;
//		 gradd[j] = 0;
	}
	JCost = 0;
	theta2 = 0;
	
	
	for(i=0; i < M; i++){
	       TX = 0;
		for(j=0; j < N; j++)
			TX = TX + theta[j]*X[i][j];
			
		g = sigmoid(TX);		
		
		JCost = JCost - y[i]*log(g) - (1.- y[i])* log(1.- g); 
		 
		for(j=0; j < N; j++){
		   grad[j] =  grad[j] + X[i][j]*(g - y[i]);		   
//		  gradd[j] = gradd[j] + g * (1. - g) * X[i][j]* X[i][j];
		   
		}	
	}
	
	
	for(j=1; j < N; j++)
		theta2= theta2 + pow(theta[j], 2);
		
	JCost = JCost/ M + lambda/2/ M * theta2 ;
	
	
		grad[0] =  grad[0]/ M;
//	     gradd[0] = gradd[0]/ M;
	for(j=1; j < N; j++){
	       grad[j] = grad[j]/ M + lambda/ M * theta[j];
//		 gradd[j] = gradd[j]/ M + lambda/ M  ;	
	}
	
	
	
	return JCost;
}
